core/deployment/src/main/java/io/quarkus/deployment/pkg/steps/JarResultBuildStep.java

```diff
@@ -415,43 +434,71 @@ private void doThinJarGeneration(CurateOutcomeBuildItem curateOutcomeBuildItem,
 
         final List<AppDependency> appDeps = curateOutcomeBuildItem.getEffectiveModel().getUserDependencies();
 
-        copyLibraryJars(transformedClasses, libDir, classPath, appDeps);
+        Map<AppDependency, List<Path>> classesOfNonJarDependencies = copyLibraryJars(transformedClasses, libDir, classPath,
+                appDeps, services);
```

Comment:
If we are doing that, why not copy them while walking the dependency tree?

Reply:
Very good point! I'll update the PR.

Reply:
Fixed